File: send-message-to-databus.py

import requests
import jwt
import time
import json
import datetime

# --- 1. PASTE YOUR NEW DATA HERE ---
api_key = 'DHm24gGQFpj3TrLLt7qqQdgtrH'
jwt_secret = 'HfbwJBfCHWzWJs7ZS0gA-gNepajAAkdT9AR-_ao2t6K'
subject_id = '0efa1b4d-95a6-4237-bd65-fdd67cae10fa'
# ------------------------------------

endpoint = 'https://dbus-qa.dataintel.xogrp.com/publish'

# --- 2. IMPROVED TOKEN GENERATION ---
print("Generating token...")

# We use datetime.UTC as recommended by the warning
now = datetime.datetime.now(datetime.UTC)

# Current time in UNIX timestamp format
now_timestamp = int(now.timestamp())

# Define validity hours
# nbf (Not Before): Valid since 60 seconds ago (to avoid clock skew)
not_before = now_timestamp - 60
# iat (Issued At): Created now
issued_at = now_timestamp
# exp (Expiration): Valid for 1 hour
expiration = now_timestamp + 3600

jwt_claims = {
    'iss': api_key,    # Issuer (Who issues it)
    'exp': expiration, # Expiration (When it expires)
    'iat': issued_at,  # Issued At (When it was created)
    'nbf': not_before  # Not Before (Valid from)
}

print(f"Token claims: {jwt_claims}")
token = jwt.encode(jwt_claims, jwt_secret, algorithm='HS256')

# --- 3. THE MESSAGE ---
timestamp = int(time.time())
stream_id = f"stream-{timestamp}"
message_body = {
    'message_id': f"message-{timestamp}",
    'description': "Test (with improved token) for the databus"
}

body_payload = {
    'subject_id': subject_id,
    'stream_id': stream_id,
    'body': message_body
}

print(f"Sending message to Subject: {subject_id}")
print(f"Content: {json.dumps(body_payload, indent=2)}")

# --- 4. PUBLICATION ---
headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {token}'
}

try:
    response = requests.post(endpoint, json=body_payload, headers=headers)
    response.raise_for_status()
    print("\nSUCCESS! (HTTP 200)")
    print("Databus accepted your message.")

except requests.exceptions.HTTPError as err:
    print(f"\nPUBLISHING ERROR! (HTTP {err.response.status_code})")
    print(f"Response: {err.response.text}")
except Exception as err:
    print(f"\nAn unexpected error occurred: {err}")

File: send-message-to-databus.js
require('dotenv').config();
const jwt = require('jsonwebtoken');
const axios = require('axios');

// --- 1. LOAD DATA FROM .ENV ---
const api_key = process.env.API_KEY;
const jwt_secret = process.env.JWT_SECRET;
const subject_id = process.env.SUBJECT_ID;

const endpoint = 'https://dbus-qa.dataintel.xogrp.com/publish';

// --- 2. IMPROVED TOKEN GENERATION ---
console.log('Generating token (v2)...');


console.log("Generating token (v2)...");
const now = new Date();
const now_timestamp = Math.floor(now.getTime() / 1000);
const not_before = now_timestamp - 60;
const issued_at = now_timestamp;
const expiration = now_timestamp + 3600;

const jwt_claims = {
    iss: api_key,
    exp: expiration,
    iat: issued_at,
    nbf: not_before
};

console.log(`Token claims: ${JSON.stringify(jwt_claims)}`);
const token = jwt.sign(jwt_claims, jwt_secret, { algorithm: 'HS256' });

// --- 3. THE MESSAGE ---
const timestamp = Math.floor(Date.now() / 1000);
const stream_id = `stream-${timestamp}`;
const message_body = {
    message_id: `msg-${timestamp}`,
    description: "Test v2 (with improved token) for the DLQ"
};

const body_payload = {
    subject_id: subject_id,
    stream_id: stream_id,
    body: message_body
};

console.log(`Sending message to Subject: ${subject_id}`);
console.log(`Content: ${JSON.stringify(body_payload, null, 2)}`);

// --- 4. PUBLICATION ---
const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${token}`
};

axios.post(endpoint, body_payload, { headers })
    .then(response => {
        console.log("\nSUCCESS! (HTTP 200)");
        console.log("Databus accepted your message.");
    })
    .catch(error => {
        if (error.response) {
            console.log(`\nPUBLISHING ERROR! (HTTP ${error.response.status})`);
            if (typeof error.response.data === 'string') {
                console.log(`Response: ${error.response.data}`);
            } else {
                console.log(`Response: ${JSON.stringify(error.response.data)}`);
            }
        } else {
            console.log(`\nAn unexpected error occurred: ${error.message}`);
        }
    });
